sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,582 @@
+    // the first column that has var-width size.
+    private int varSizeColIndex;
+    // At what point validity data starts.
+    private int validityBytesOffset;
+    // Caches the estimate size (8-bytes aligned). Calculating and reading this field is
+    // synchronized.
+    private int estimateRowSize;
+
+    private RowOffsetsCalculator(Attribute[] attrArr, int[] offsetArr) {
+      // offsetArr can be null.
+      // This is in the context of constructing an object to get an estimate of the memory needed
+      // by the row.
+      this.attributes = attrArr;
+      resetMeta();
+      initFixedWidthSection(offsetArr);
+    }
+
+    /**
+     * Handles iterating on all the columns to advance the {@link #byteCursor}.
+     * Then, it calculates the start offset of the validity and its size in bytes.
+     * This method updates the following: {@link #byteCursor}, {@link #offsets},
+     * {@link #varSizeColIndex}, and all the fields related to the validity
+     *
+     * @param offsetArr if not NULL, it will be used to save the column offsets.
+     */
+    private void initFixedWidthSection(int[] offsetArr) {
+      if (offsetArr == null) {
+        for (int i = 0; i < attributes.length; i++) {
+          advanceColCursorAndGet(i);
+        }
+      } else {
+        this.offsets = offsetArr;
+        for (int i = 0; i < attributes.length; i++) {
+          offsets[i] = advanceColCursorAndGet(i);
+        }
+      }
+      // set the validity size, and mark where it starts
+      setValidityBits(attributes.length);
+    }
+
+    /**
+     * Given the column-index, it increments the {@link #byteCursor} by the length of the datatype
+     * of that column.
+     * It takes into considerations the alignment of the column.
+     * Note that this method expects sequential increments to the column. Accessing columns in
+     * random order would result in incorrect calculations.
+     *
+     * @param ind the index of the column in the schema.
+     * @return At what point the column starts.
+     */
+    private int advanceColCursorAndGet(int ind) {
+      int res;
+      Attribute attr = attributes[ind];
+      DataType dataType = attr.dataType();
+      DType rapidsType = GpuColumnVector.getNonNestedRapidsType(dataType);
+      if (isFixedLength(dataType)) {
+        int length = rapidsType.getSizeInBytes();
+        res = alignOffset(byteCursor, length);
+        byteCursor = res + length;
+        return res;
+      }
+      if (!hasVarSizeData()) {
+        // set the index to the first variable size column.
+        varSizeColIndex = ind;
+      }
+      res = alignOffset(byteCursor, JCUDF_VAR_LENGTH_FIELD_ALIGNMENT);
+      byteCursor = res + JCUDF_VAR_LENGTH_FIELD_SIZE;
+      return res;
+    }
+
+    private void resetMeta() {
+      varSizeColIndex = attributes.length;
+      byteCursor = 0;
+      estimateRowSize = -1;
+      validityBytesOffset = 0;
+    }
+
+    private void setValidityBits(int fieldsCount) {
```

Comment:
I had to double-take this because I was looking for the validity data is was setting. Maybe better named something like `addValiditySizeToByteCursor`?